### DISCUSSION TITLE
Recommendations for setting Reactor.p.time for calendar time?

### DISCUSSION THREAD

**Discussion by drewj-usnctech** [terrapower/armi#1475] [Q&A] [answered]
I can't seem to find anything inside the ARMI framework that sets the calendar time for a simulation. There's this `Reactor.p.time` parameter
https://github.com/terrapower/armi/blob/6603c1daac9e948d55b8f1a93245dec7192d9b41/armi/reactor/reactorParameters.py#L78-L83
with a depletion description

but it doesn't seem to be set inside ARMI. It looks like the cycle and node integers are serialized, but then _something_ must set `Reactor.p.time` in order for it to be useful in e.g., report generation
https://github.com/terrapower/armi/blob/6603c1daac9e948d55b8f1a93245dec7192d9b41/armi/physics/neutronics/reports.py#L98

And it seems to me that this is something that the main `Operator` or a shipped interface could handle relatively easily? The `Operator` handles marching from `(cycle, node)` to `(cycle, node)`, so it could conceivably also set `Reactor.p.time`

We have a kind of complex routine that re-constructs the time from the `(cycle, node)` value at `interactEveryNode` and it's got me thinking there's got to be a better way.

So the question(s) are

1. What is the recommended way to set `Reactor.p.time` in a depletion rune?
2. Is this something that ARMI could standardize such that the field is set as the simulation marches in time, prior to calls to `Interface.interact*`?

**Comment by keckler** [comment]
@drewj-usnctech Your analysis is spot-on. Check out this issue, it has some context that probably answers your question: https://github.com/terrapower/armi/issues/1100

Long story short, right now this time parameter is advanced by our internal depletion plugin. I don't think there is any particularly good reason for that.

**Reply by drewj-usnctech** [reply]
Thanks @keckler. Question isn't really answered, but I've got more threads to pull now

**Reply by john-science** [reply]
@keckler At some point, I would like to pick your brain about this.

**Comment by drewj-tp** [comment]
@john-science and I are looking into this

**Reply by john-science** [reply]
Notes.

That comes from here:
 
https://github.com/terrapower/armi/blob/ee691abf61d56197515886ec5bbb4804b4673a90/armi/utils/__init__.py#L178
 
The ARMI settings in questions:  `nCycles`, `cycleLengths`. Both defined in ARMI.  Though I think the setting step days is sometimes used.

For example: https://github.com/terrapower/armi/blob/ee691abf61d56197515886ec5bbb4804b4673a90/armi/tests/anl-afci-177/anl-afci-177.yaml#L9-L15

**Reply by john-science** [reply]
I think this requires a two-step approach:

1. Move the one line to set `r.p.time` into ARMI (at the VERY beginning of the time step).
2. Document this somewhere in the ARMI docs.

The (1) here will require some downstream testing.

**Reply by drewj-tp** [reply]
I think that's adequate. My preference would be that this is something the operator does before the interface stack restarts. That is, *every* interface knows `r.p.time` is set correctly

**Reply by john-science** [reply]
Okay, that's a fair improvement.  We already do things like that in `Operator`:

https://github.com/terrapower/armi/blob/3431c44859a88ead124dfd323781c70c39bae714/armi/operators/operator.py#L394

https://github.com/terrapower/armi/blob/3431c44859a88ead124dfd323781c70c39bae714/armi/operators/operator.py#L427-L431

Seems like we could find a good place for that.

**Comment by drewj-tp** [comment]
Fun little easter egg: if some later plugin is responsible for updating the `Reactor.p.time` parameter, your logs will look out of sync
```
=========== 01 - main                        EveryNode - timestep: cycle 0, node 1, year 0.00 ===========
=========== XX - depletion                   EveryNode - timestep: cycle 0, node 1, year 0.00 ===========
=========== YY - report                      EveryNode - timestep: cycle 0, node 1, year 0.75 ===========
=========== ZZ - database                    EveryNode - timestep: cycle 0, node 1, year 0.75 ===========
```
for `XX < YY < ZZ` in the interface stack

**Comment by drewj-tp** [comment, marked as answer]
https://github.com/terrapower/armi/pull/2225 is handling this
